File: leetcode/BinarySearch/leetcode-81/81.java
class Solution {
    public boolean search(int[] nums, int target) {
        if (nums == null || nums.length == 0) {
            return false;
        }

        int left = 0, right = nums.length - 1;
        while (left < right - 1) {
            int mid = (right - left) / 2 + left;
            if (nums[mid] == target) {
                return true;
            } else if (nums[left] < nums[mid]) {
                if (nums[left] <= target && nums[mid] > target) {
                    right = mid;
                } else {
                    left = mid;
                }
            } else if (nums[left] > nums[mid]) {
                if (nums[mid] < target && nums[right] >= target) {
                    left = mid;
                } else {
                    right = mid;
                }
            } else {
                left++;
            }
        }
        if (nums[left] == target) {
            return true;
        }
        if (nums[right] == target) {
            return true;
        }
        return false;
    }
}

File: leetcode/BinarySearch/leetcode-81/81.go
package binarysearch

func search(nums []int, target int) bool {
	if nums == nil || len(nums) == 0 {
		return false
	}

	start, end := 0, len(nums)-1
	for start < end-1 {
		mid := (end-start)/2 + start
		if nums[mid] == target {
			return true
		} else if nums[start] < nums[mid] {
			if nums[start] <= target && target < nums[mid] {
				end = mid
			} else {
				start = mid
			}
		} else if nums[start] > nums[mid] {
			if nums[mid] < target && target <= nums[end] {
				start = mid
			} else {
				end = mid
			}
		} else {
			start++
		}
	}

	if nums[start] == target {
		return true
	}
	if nums[end] == target {
		return true
	}
	return false
}
